### PR TITLE
Add X-Original-URI support

### LINKED ISSUE #527: Mounting flask under subpath through nginx
Sorry if this is obvious and I just missed how to do it. My flask app is the rest API so http://localhost:5000/swagger.json is correct. However, if I mount it under /api via a proxy_pass in nginx, the swagger ui is still looking for /swagger.json instead of /api/swagger.json

Setting base_path relocates the whole api so I end up with /api/api/my-rest-call once mounted in nginx.

Is there a way to specify the path where the swagger.json file is located? For now I created a rewrite rule on nginx but that is not a long term solution in case multiple Connexion apps are running under the same nginx instance.

### PR DESCRIPTION
**Pull request by Jyhess**
Fixes #527 .

Changes proposed in this pull request:

 - Dynamically update base_base using X-Original-URI header to support connexion behind a proxy

### CODE REVIEW

**Comment by Jyhess** [comment]
@dtkav I had a look on your PR, but didn't understand how you succeed to get path part modified by the reverse proxy. Can you give me a hint ?

Using an header make a solution generic, dynamic and configurable for all modern reverse proxy. But I agree with you the header name should be configurable and documented.
For a more static solution, we can use an application option like `root_uri`. It can fix this issue, and is easier to understand. But it is less dynamic, and will not work if same application can be accessed through 2 different reverse proxy. Even if its not common and maybe not a good practice, I had this case in the past with 2 reverse proxies (for admins and users) using 2 different paths.

**Comment by Jyhess** [comment]
@dtkav Looking example you added in your last commit, I understand we are not trying to fix the same issue.
If I well understand, you are fixing the case when application is mounted under a subpath in a WSGI server. Like it is well explained in #392.
On my side, I'm trying to solve the same issue, but when a reverse proxy is used as a completely separated service with rewrite rules:
http://mydomain.com/api/myservice/v1/index.html -> reverse proxy service -> http://myservice/v1/index.html -> my service
In this case, both services are running in different process, (different docker on my case). So there is no way for myservice to know where it is _mounted_. To fix this case, there is mainly 2 solutions:
- A static configuration for _my service_ used to define proxied base path. Easy to implement, but it introduce a configuration coupling between 2 services, and not allow to have same service mounted under 2 different path (I add this case in a past where the same service was used under /api/ and /admin/)
- Configure the reverse proxy to forward original hostname and path via HTTP header. This is the solution I'm proposing.

By the way, this will not solve #527 as I guessed originally. Do you think I should open a different issue with my use case explained ?

**Comment by Jyhess** [comment]
I replaced cookbook minimal example with a more complete one in `examples/openapi3/external_reverse_proxy`

**Comment by dtkav** [comment]
@Jyhess thanks for laying out the issue so clearly.

Please have another look at the example I posted - I forgot to add the file where all the magic was happening :)

I believe it supports your use case by passing the header `X-Script-Name`

Let me know if I'm still missing something!

**Comment by dtkav** [comment]
@Jyhess my mistake - I didn't realize your use case was with aiohttp (which I'm not very familiar with).
I will read more about url_for for aiohttp. There must be a way to do this with the framework - it's such a common use case to generate fully qualified urls within a web framework.

> What I can do, is to port header based solution for AioHttp in your branch, with header name alignment, and keep your solution for Flask. Do you agree with this plan ?

I don't think `connexion/apis/*` should be proxy-aware. The reason I'm going down the url_for path, is that I think that is the right layer of abstraction to fix it.

Many users will be using reverse proxies, and they will often need to rely on this type of functionality (access to incoming path, scheme, host). If it is implemented with the primitives available from the underlying web framework, then it's available in the handler functions. If the user understands the framework well, they won't have to learn anything new.

> Note your ReverseProxied class is not usable with AioHttp. WSGI native support has been removed and is now part of an external third party. So if you introduce a WSGI module in connexion and want to keep Flask and AioHttp compatibility, you will need more works. Maybe it can be easier to move your wsgi.py in your example folder instead of main source code.

Ok, that makes sense. I didn't even realise that aiohttp had a wsgi compatibility layer at one point.
I agree that it probably makes sense to move wsgi.py into the example folder, or potentially built into `connexion/apps/flask_app.py`.

For aiohttp, I think the best way to do this is to set the `original uri`/`script_name` in the middleware along with the scheme and host. I've just started reading about this, but here's what I found so far:
https://github.com/aio-libs/aiohttp-remotes/blob/master/aiohttp_remotes/x_forwarded.py
http://docs.aiohttp.org/en/stable/web_advanced.html#deploying-behind-a-proxy

So, to sum up:
1) I don't think reverse-proxy logic belongs in `connexion/apis/*`
2) I think we should rely on the primitives of the underlying framework
3) I will learn more about aiohttp, and try to come up with a solution using aiohttp middlewear

I appreciate your clear communication style. Hopefully I'm making sense too :grin: 

Let me know what you think!

**Comment by dtkav** [comment]
I think I got it working with aiohttp in #823 today.
I'd love your feedback now that they are more concretely comparable.

**Comment by Jyhess** [comment]
Good catch for aiohttp-remotes. I missed it during my own research.
This solution is better than mine. I will do some minor comments on your PR.
Thanks for work.

**Comment by dtkav** [comment]
Thanks for the collaboration @Jyhess !

**Comment by Jyhess** [comment]
Thanks to you Daniel.
